### PR TITLE
Fix type instability in unflattening Tuples

### PR DESCRIPTION
**Pull request by simsurace**
On master this does not infer:
```julia
julia> tpl = (1., ((2., 3.), 4.));
julia> par, unflatten = flatten(tpl);
julia> @code_warntype unflatten(par)
```
This PR fixes this and (so I believe) improves the type inference with a recursive implementation and a fallback for empty tuples.

### CODE REVIEW

File: test/flatten.jl [hunk 1]
```diff
@@ -39,6 +39,13 @@
         test_flatten_interface((1.0, 2.0); check_inferred=tuple_infers)
 
         test_flatten_interface((1.0, (2.0, 3.0), randn(5)); check_inferred=tuple_infers)
+
+        # Prevent regression of PR #67
+        @testset "Type stability of unflatten" begin
+            θ = (1., ((2., 3.), 4.))
```
[the remark this answers is not in the document]
Comment (simsurace) [reply test/flatten.jl:45]: This dog is so picky :)

**Comment by willtebbutt** [comment]
Thanks for opening this. Busy morning -- will try to review later today

File: src/flatten.jl [hunk 2]
```diff
@@ -79,16 +79,20 @@ function flatten(::Type{T}, x::SparseMatrixCSC) where {T<:Real}
...
-    sz = _cumsum(lengths)
+    vec1, back1 = flatten(T, first(x))
+    vec2, back2 = flatten(T, Base.tail(x))
+    l1 = length(vec1)
+    l2 = length(vec2)
     function unflatten_to_Tuple(v::Vector{T})
-        map(x_backs, lengths, sz) do x_back, l, s
-            return x_back(v[(s - l + 1):s])
-        end
+        return (back1(v[1:l1]), back2(v[(l1 + 1):(l1 + l2)])...)
```
Comment (simsurace) [inline src/flatten.jl:87]: Maybe it would be better to use views here?

**Review by simsurace — COMMENTED** [review]
What is your view on using `view`s (pun intended)?

**Comment by willtebbutt** [comment]
I think views definitely make sense here. Very open to that.

While you do that, I'm going to drop 1.5, because we really don't need it.

**Comment by simsurace** [comment]
Thinking it over a bit, I think going to views should be evaluated a bit more carefully, as it also requires generalizing a lot of the signatures of the `unflatten` functions. I will keep that for a later PR perhaps.

File: src/flatten.jl (hunk 2 above)
Comment (simsurace) [reply src/flatten.jl:87]: Let's keep that for later.

**Comment by simsurace** [comment]
I'm gonna suggest approving this as is. I bumped the version. Maybe we can release #68 in the same swoop?

**Comment by willtebbutt** [comment]
> Thinking it over a bit, I think going to views should be evaluated a bit more carefully, as it also requires generalizing a lot of the signatures of the unflatten functions. I will keep that for a later PR perhaps.

Good point. Yeah, let's defer that until later. I do think it would be good to look at though, because we're currently doing a _lot_ of copying.

**Review by willtebbutt — APPROVED** [review]
LGTM.

I'd really quite like to stick to the standard 1-version-per-PR rule though, so please tag this separately from #68

**Comment by simsurace** [comment]
No problem!

**Pull request merged**